pkg/apiserver/authorization/authorizerfactory/opa.go

```diff
@@ -33,125 +34,140 @@ type opaAuthorizer struct {
-		return authorizer.DecisionDeny, permissionUndefined, nil
+	// it's global resource, permission denied
+	if attr.GetResourceScope() == iamv1alpha2.GlobalScope {
+		return authorizer.DecisionNoOpinion, "", nil
 	}
 
-	clusterRole, err := o.am.GetRoleOfUserInTargetScope(iamv1alpha2.ClusterScope, attr.GetCluster(), attr.GetUser().GetName())
-	if err != nil {
-		if errors.IsNotFound(err) {
-			return authorizer.DecisionDeny, err.Error(), nil
+	if attr.GetResourceScope() == iamv1alpha2.WorkspaceScope {
+		workspaceRole, err := o.am.GetWorkspaceRoleOfUser(attr.GetUser().GetName(), attr.GetWorkspace())
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return authorizer.DecisionNoOpinion, "", nil
+			}
+			return authorizer.DecisionNoOpinion, "", err
 		}
-		return authorizer.DecisionDeny, "", err
-	}
 
-	// check cluster role policy rules
-	if authorized, reason, err := o.makeDecision(clusterRole, attr); authorized == authorizer.DecisionAllow {
-		return authorized, reason, nil
-	} else if err != nil {
-		return authorizer.DecisionDeny, "", err
-	}
+		// check workspace role policy rules
+		if authorized, reason, err := o.makeDecision(workspaceRole, attr); authorized == authorizer.DecisionAllow {
+			return authorized, reason, err
+		} else if err != nil {
+			return authorizer.DecisionNoOpinion, "", err
+		}
 
-	// it's not in cluster resource, permission denied
-	if attr.GetWorkspace() == "" && attr.GetNamespace() == "" {
-		return authorizer.DecisionDeny, permissionUndefined, nil
+		return authorizer.DecisionNoOpinion, "", nil
 	}
 
-	workspaceRole, err := o.am.GetRoleOfUserInTargetScope(iamv1alpha2.WorkspaceScope, attr.GetWorkspace(), attr.GetUser().GetName())
-	if err != nil {
-		if errors.IsNotFound(err) {
-			return authorizer.DecisionDeny, err.Error(), nil
+	if attr.GetResourceScope() == iamv1alpha2.NamespaceScope {
+		role, err := o.am.GetNamespaceRoleOfUser(attr.GetUser().GetName(), attr.GetNamespace())
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return authorizer.DecisionNoOpinion, "", nil
+			}
+			return authorizer.DecisionNoOpinion, "", err
+		}
+		// check namespace role policy rules
+		if authorized, reason, err := o.makeDecision(role, attr); authorized == authorizer.DecisionAllow {
+			return authorized, reason, err
+		} else if err != nil {
+			return authorizer.DecisionNoOpinion, "", err
 		}
-		return authorizer.DecisionDeny, "", err
-	}
 
-	// check workspace role policy rules
-	if authorized, reason, err := o.makeDecision(workspaceRole, attr); authorized == authorizer.DecisionAllow {
-		return authorized, reason, err
-	} else if err != nil {
-		return authorizer.DecisionDeny, "", err
+		return authorizer.DecisionNoOpinion, "", nil
 	}
 
-	// it's not in workspace resource, permission denied
-	if attr.GetNamespace() == "" {
-		return authorizer.DecisionDeny, permissionUndefined, nil
+	clusterRole, err := o.am.GetClusterRoleOfUser(attr.GetUser().GetName(), attr.GetCluster())
+
+	if errors.IsNotFound(err) {
```

Comment:
check err != nil first

Reply:
It's unnecessary, if err is nil, always false.